File: leetcode_solved/leetcode_0543_Diameter_of_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
    int maxdiadepth = 0;
    int dfs(TreeNode *root)
    {
        if (root == NULL)
            return 0;
        int leftdepth = dfs(root->left);
        int rightdepth = dfs(root->right);
        if (leftdepth + rightdepth > maxdiadepth)
            maxdiadepth = leftdepth + rightdepth;
        return max(leftdepth + 1, rightdepth + 1);
    }
    int diameterOfBinaryTree(TreeNode *root)
    {
        dfs(root);
        return maxdiadepth;
    }
};

File: leetcode_solved/leetcode_0543_Diameter_of_Binary_Tree.java
// AC: Runtime: 0 ms, faster than 100.00% of Java online submissions for Diameter of Binary Tree.
// Memory Usage: 38.9 MB, less than 67.84% of Java online submissions for Diameter of Binary Tree.
// recursive. remain the max depth of left child tree and right child tree, and max(left, right) + 1 return to upper parent node.
// T:O(n), S:O(logn)
//
class Solution {
    private int MAX_DIS = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        preOrderTraversal(root);
        return MAX_DIS;
    }

    private int preOrderTraversal(TreeNode root) {
        if (root == null) {
            return 0;
        }
        if (root.left == null && root.right == null) {
            return 1;
        }

        int left = 0, right = 0;
        if (root.left != null) {
            left = preOrderTraversal(root.left);
        }
        if (root.right != null) {
            right = preOrderTraversal(root.right);
        }

        MAX_DIS = Math.max(MAX_DIS, left + right);

        return Math.max(left, right) + 1;
    }
}